### PR TITLE
Modify API for advanced settings (RandomizedPcaTrainer)

### LINKED ISSUE #2471: Missing support for Anomaly Detection metrics.
ML.NET supports several evaluation metrics but is missing support for Anomaly Detection metrics.

Example of supported metrics:
- [BinaryClassificationMetrics](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Evaluators/Metrics/BinaryClassificationMetrics.cs)
- [RegressionMetrics](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Evaluators/Metrics/RegressionMetrics.cs)

Is anomaly detection a scenario we want to support for V 1.0 ?  If yes, we should probably add support for anomaly detection metrics

@Ivanidzo4ka @sfilipi @shauheen @glebuk @artidoro

### PR DESCRIPTION
**Pull request by abgoswam**
Towards #1798 . 
+ adds MLContext extension for Anomaly detection Issue #1369
+ fixes #2471 

This PR addresses the following algos

- RandomizedPcaTrainer

The following changes have been made:

- Make constructors internal .
- Rename Arguments to Options
- Rename Options objects as options (instead of args or advancedSettings used so far)

### CODE REVIEW

File: src/Microsoft.ML.PCA/PcaTrainer.cs [hunk 1]
```diff
@@ -103,23 +103,23 @@ public class Arguments : UnsupervisedLearnerInputBaseWithWeight
 
         }
 
-        internal RandomizedPcaTrainer(IHostEnvironment env, Arguments args)
-            :this(env, args, args.FeatureColumn, args.WeightColumn)
+        internal RandomizedPcaTrainer(IHostEnvironment env, Options options)
```
Comment (artidoro) [inline src/Microsoft.ML.PCA/PcaTrainer.cs]: >RandomizedPcaTrainer [](start = 17, length = 20)

It's strange... I noticed that renaming Arguments to Options did not modify anything in the mlContext catalog.

Comment (artidoro) [reply src/Microsoft.ML.PCA/PcaTrainer.cs]: I looked it up, and I don't think there is an entry for this trainer in mlContext. Can you add it?

---
In reply to: [253319239](https://github.com/dotnet/machinelearning/pull/2390#discussion_r253319239) [](ancestors = 253319239)

Comment (abgoswam) [reply src/Microsoft.ML.PCA/PcaTrainer.cs]: yeah.  i noticed couple more components which do not have mlcontext extension. 

will add

---
In reply to: [253319255](https://github.com/dotnet/machinelearning/pull/2390#discussion_r253319255) [](ancestors = 253319255,253319239)

File: src/Microsoft.ML.PCA/PcaTrainer.cs [hunk 2]
```diff
@@ -49,7 +49,7 @@ public sealed class RandomizedPcaTrainer : TrainerEstimatorBase<AnomalyPredictio
         internal const string Summary = "This algorithm trains an approximate PCA using Randomized SVD algorithm. "
             + "This PCA can be made into Kernel PCA by using Random Fourier Features transform.";
 
-        public class Arguments : UnsupervisedLearnerInputBaseWithWeight
+        public class Options : UnsupervisedLearnerInputBaseWithWeight
```
Comment (sfilipi) [inline src/Microsoft.ML.PCA/PcaTrainer.cs]: >Options [](start = 21, length = 7)

xml docs are coming later?

File: src/Microsoft.ML.PCA/PcaTrainer.cs [hunk 3]
```diff
@@ -91,7 +91,7 @@ public class Arguments : UnsupervisedLearnerInputBaseWithWeight
         /// <param name="oversampling">Oversampling parameter for randomized PCA training.</param>
         /// <param name="center">If enabled, data is centered to be zero mean.</param>
         /// <param name="seed">The seed for random number generation.</param>
-        public RandomizedPcaTrainer(IHostEnvironment env,
+        internal RandomizedPcaTrainer(IHostEnvironment env,
```
Comment (rogancarr) [inline src/Microsoft.ML.PCA/PcaTrainer.cs]: Shouldn't we just make the class `Internal`/`BestFriend` and keep this `public`?

File: src/Microsoft.ML.PCA/PcaTrainer.cs [hunk 4]
```diff
@@ -347,14 +347,14 @@ protected override AnomalyPredictionTransformer<PcaModelParameters> MakeTransfor
             Desc = "Train an PCA Anomaly model.",
             UserName = UserNameValue,
             ShortName = ShortName)]
-        internal static CommonOutputs.AnomalyDetectionOutput TrainPcaAnomaly(IHostEnvironment env, Arguments input)
+        internal static CommonOutputs.AnomalyDetectionOutput TrainPcaAnomaly(IHostEnvironment env, Options input)
```
Comment (rogancarr) [inline src/Microsoft.ML.PCA/PcaTrainer.cs]: Same as above; these can be kept `public` and the whole class can be made `internal`/`BestFriend`.

File: src/Microsoft.ML.PCA/PcaTrainer.cs (hunk 3 above)
Comment (abgoswam) [reply src/Microsoft.ML.PCA/PcaTrainer.cs]: not really. 

we would want to expose these through mlcontext. not via *constructors*

---
In reply to: [254391018](https://github.com/dotnet/machinelearning/pull/2390#discussion_r254391018) [](ancestors = 254391018)

File: src/Microsoft.ML.PCA/PcaTrainer.cs (hunk 4 above)
Comment (abgoswam) [reply src/Microsoft.ML.PCA/PcaTrainer.cs]: The `Options` class above should be public.  Hence we cannot make entire class internal.

---
In reply to: [254391534](https://github.com/dotnet/machinelearning/pull/2390#discussion_r254391534) [](ancestors = 254391534)

File: src/Microsoft.ML.PCA/PcaTrainer.cs (hunk 2 above)
Comment (abgoswam) [reply src/Microsoft.ML.PCA/PcaTrainer.cs]: yeap. 

---
In reply to: [254353429](https://github.com/dotnet/machinelearning/pull/2390#discussion_r254353429) [](ancestors = 254353429)

File: src/Microsoft.ML.PCA/PcaTrainer.cs (hunk 1 above)
Comment (abgoswam) [reply src/Microsoft.ML.PCA/PcaTrainer.cs]: i added mlcontext  extension for this.  Also added a test  for it that exercises the Fit() and Transform()  APIs.

 Evaluate() API currently missing from Anomaly Detection. i will create a separate issue for that.

---
In reply to: [253584603](https://github.com/dotnet/machinelearning/pull/2390#discussion_r253584603) [](ancestors = 253584603,253319255,253319239)

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+                Columns = new[]
+                {
+                    new TextLoader.Column(featureColumn, DataKind.R4, new [] { new TextLoader.Range(1, 784) })
+                }
+            });
+
+            var trainData = reader.Read(GetDataPath(TestDatasets.mnistOneClass.trainFilename));
+            var testData = reader.Read(GetDataPath(TestDatasets.mnistOneClass.testFilename));
+
+            var pipeline = mlContext.AnomalyDetection.Trainers.RandomizedPca(featureColumn);
```
Comment (abgoswam) [inline test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: >AnomalyDetection [](start = 37, length = 16)

Anomaly detection currently does not support evaluation metrics 

Issue #2471

File: src/Microsoft.ML.Data/Evaluators/AnomalyDetectionEvaluator.cs [hunk 6]
```diff
@@ -576,6 +577,43 @@ public void Finish()
...
+
+        /// <summary>
+        /// Evaluates scored anomaly detection data.
+        /// </summary>
+        /// <param name="data">The scored data.</param>
+        /// <param name="label">The name of the label column in <paramref name="data"/>.</param>
+        /// <param name="score">The name of the score column in <paramref name="data"/>.</param>
+        /// <param name="predictedLabel">The name of the predicted label column in <paramref name="data"/>.</param>
+        /// <returns>The evaluation results for these outputs.</returns>
+        public AnomalyDetectionMetrics Evaluate(IDataView data, string label, string score, string predictedLabel)
```
Comment (rogancarr) [inline src/Microsoft.ML.Data/Evaluators/AnomalyDetectionEvaluator.cs]: Defaults for `label`, `score`, and `predictedLabel`?

File: src/Microsoft.ML.Data/Evaluators/Metrics/AnomalyDetectionMetrics.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Microsoft.Data.DataView;
+
+namespace Microsoft.ML.Data.Evaluators.Metrics
+{
+    public sealed class AnomalyDetectionMetrics
+    {
+        public double Auc { get; }
```
Comment (rogancarr) [inline src/Microsoft.ML.Data/Evaluators/Metrics/AnomalyDetectionMetrics.cs]: Summaries, Remarks, and links to relevant documentation.

File: src/Microsoft.ML.PCA/PCACatalog.cs [hunk 8]
```diff
@@ -35,5 +36,25 @@ public static class PcaCatalog
         /// <param name="columns">Input columns to apply PrincipalComponentAnalysis on.</param>
         public static PrincipalComponentAnalysisEstimator ProjectToPrincipalComponents(this TransformsCatalog.ProjectionTransforms catalog, params PrincipalComponentAnalysisEstimator.ColumnInfo[] columns)
             => new PrincipalComponentAnalysisEstimator(CatalogUtils.GetEnvironment(catalog), columns);
+
+        public static RandomizedPcaTrainer RandomizedPca(this AnomalyDetectionCatalog.AnomalyDetectionTrainers catalog,
```
Comment (rogancarr) [inline src/Microsoft.ML.PCA/PCACatalog.cs]: Needs xml docs with remarks and links to a sample. Here or add to #1209 .

File: src/Microsoft.ML.PCA/PcaTrainer.cs (hunk 3 above)
Comment (rogancarr) [reply src/Microsoft.ML.PCA/PcaTrainer.cs]: Resolved. I hadn't seen the pattern for trainable transforms where the class is `public` and methods are `internal`.

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+            var pipeline = ML.AnomalyDetection.Trainers.RandomizedPca(featureColumn);
+
+            var transformer = pipeline.Fit(trainData);
+            var transformedData = transformer.Transform(testData);
+
+            // Evaluate
+            var metrics = ML.AnomalyDetection.Evaluate(transformedData);
+
+            Assert.Equal(0.99, metrics.Auc, 2);
```
Comment (rogancarr) [inline test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: How do we know that these numbers are correct?

Comment (abgoswam) [reply test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: i tried out the same dataset in TLC with the same trainer, the numbers are *close*.

in general, in this PR i am only exposing the  trainer / evaluators  as they exist  currently in the codebase.  the PR does not have  any algorithmic changes or  changes in evaluation metrics themselves.

---
In reply to: [255589211](https://github.com/dotnet/machinelearning/pull/2390#discussion_r255589211) [](ancestors = 255589211)

File: src/Microsoft.ML.PCA/PcaTrainer.cs (hunk 3 above)
Comment (abgoswam) [reply src/Microsoft.ML.PCA/PcaTrainer.cs]: in ML.NET  terms,  this is "trainer estimator"  for anomaly detection tasks.  most other  "trainer estimator"s follow the same pattern  e.g.  KMeansPlusPlusTrainer

---
In reply to: [255587702](https://github.com/dotnet/machinelearning/pull/2390#discussion_r255587702) [](ancestors = 255587702)

File: src/Microsoft.ML.PCA/PCACatalog.cs (hunk 8 above)
Comment (abgoswam) [reply src/Microsoft.ML.PCA/PCACatalog.cs]: added xml docs.

sample adding can be part of overall documentation effort #1209

---
In reply to: [255585449](https://github.com/dotnet/machinelearning/pull/2390#discussion_r255585449) [](ancestors = 255585449)

File: src/Microsoft.ML.Data/Evaluators/Metrics/AnomalyDetectionMetrics.cs (hunk 7 above)
Comment (abgoswam) [reply src/Microsoft.ML.Data/Evaluators/Metrics/AnomalyDetectionMetrics.cs]: added basic summaries for now.

wanted to also add the remarks from TLC website., but the explanations there were not clear esp. for the detection rate metrics.  

---
In reply to: [255583277](https://github.com/dotnet/machinelearning/pull/2390#discussion_r255583277) [](ancestors = 255583277)

File: src/Microsoft.ML.Data/TrainCatalog.cs [hunk 10]
```diff
@@ -564,4 +565,51 @@ public RankerMetrics Evaluate(IDataView data, string label, string groupId, stri
...
+        }
+
+        /// <summary>
+        /// Evaluates scored anomaly detection data.
+        /// </summary>
+        /// <param name="data">The scored data.</param>
+        /// <param name="label">The name of the label column in <paramref name="data"/>.</param>
+        /// <param name="score">The name of the score column in <paramref name="data"/>.</param>
+        /// <param name="predictedLabel">The name of the predicted label column in <paramref name="data"/>.</param>
+        /// <returns>The evaluation results for these calibrated outputs.</returns>
```
Comment (artidoro) [inline src/Microsoft.ML.Data/TrainCatalog.cs]: >calibrated [](start = 54, length = 10)

What is calibrated here? Could you explain in a few more words?

File: src/Microsoft.ML.Data/Evaluators/Metrics/AnomalyDetectionMetrics.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        /// (assuming 'positive' ranks higher than 'negative').
+        /// </remarks>
+        public double Auc { get; }
+
+        /// <summary>
+        /// Detection rate at k false positives.
+        /// </summary>
+        public double DrAtK { get; }
+        /// <summary>
+        /// Detection rate at fraction p false positives.
```
Comment (artidoro) [inline src/Microsoft.ML.Data/Evaluators/Metrics/AnomalyDetectionMetrics.cs]: >p [](start = 39, length = 1)

What is `p`?

Comment (abgoswam) [reply src/Microsoft.ML.Data/Evaluators/Metrics/AnomalyDetectionMetrics.cs]: to me the Detection rate metrics look suspicious.  perhaps we should not expose these for V1.0 .  only expose AUC

---
In reply to: [255719011](https://github.com/dotnet/machinelearning/pull/2390#discussion_r255719011) [](ancestors = 255719011)

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs (hunk 9 above)
Comment (rogancarr) [reply test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: Can we mark that this is a baseline test, then, and not a functionality test?

---
In reply to: [255680814](https://github.com/dotnet/machinelearning/pull/2390#discussion_r255680814) [](ancestors = 255680814,255589211)

**Comment by rogancarr** [comment]
Just one last question about tests!

File: src/Microsoft.ML.Data/TrainCatalog.cs (hunk 10 above)
Comment (abgoswam) [reply src/Microsoft.ML.Data/TrainCatalog.cs]: it was a copy-paste typo. fixed... 

---
In reply to: [255717912](https://github.com/dotnet/machinelearning/pull/2390#discussion_r255717912) [](ancestors = 255717912)

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs (hunk 9 above)
Comment (abgoswam) [reply test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: good point. 

as per the classification above, these seem like "baseline" tests, I have increased the precision to 5 places of decimal.

as for test migration from internal repo, it seems we ported over the `PcaAnomalyTest`  already  . that should suffice for correctness.

---
In reply to: [255796608](https://github.com/dotnet/machinelearning/pull/2390#discussion_r255796608) [](ancestors = 255796608,255680814,255589211)

File: src/Microsoft.ML.Data/Evaluators/Metrics/AnomalyDetectionMetrics.cs (hunk 11 above)
Comment (abgoswam) [reply src/Microsoft.ML.Data/Evaluators/Metrics/AnomalyDetectionMetrics.cs]: removed that detectionrate@p metric.

---
In reply to: [255725823](https://github.com/dotnet/machinelearning/pull/2390#discussion_r255725823) [](ancestors = 255725823,255719011)

File: src/Microsoft.ML.Data/Evaluators/Metrics/AnomalyDetectionMetrics.cs (hunk 7 above)
Comment (rogancarr) [reply src/Microsoft.ML.Data/Evaluators/Metrics/AnomalyDetectionMetrics.cs]: For these summaries, check in with @shmoradims ; he's building a set of generic docs for things like AUC, F1, RMSE, etc.

---
In reply to: [255703503](https://github.com/dotnet/machinelearning/pull/2390#discussion_r255703503) [](ancestors = 255703503,255583277)

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+            var pipeline = ML.AnomalyDetection.Trainers.RandomizedPca(featureColumn);
+
+            var transformer = pipeline.Fit(trainData);
+            var transformedData = transformer.Transform(testData);
+
+            // Evaluate
+            var metrics = ML.AnomalyDetection.Evaluate(transformedData, k: 10);
+
+            Assert.Equal(0.98558, metrics.Auc, 5);
+            Assert.Equal(0.90, metrics.DrAtK, 2);
```
Comment (rogancarr) [inline test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: > Assert.Equal(0.90, metrics.DrAtK, 2); [](start = 11, length = 38)

This one too, please :)

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        public AnomalyDetectionTests(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        /// <summary>
+        /// RandomizedPcaTrainer test 
+        /// </summary>
+        [Fact]
+        public void RandomizedPcaTrainer()
```
Comment (rogancarr) [inline test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: >RandomizedPcaTrainer [](start = 20, length = 20)

`RandomizedPcaTrainerBaselineTest`

**Comment by abgoswam** [comment]
Thanks for the reviews!
